### PR TITLE
add error context for rate limit event

### PR DESCRIPTION
**Pull request by dillonstreator**
In order to provide better insight around context deadline exceeded events, this PR adds error context.

### CODE REVIEW

File: athenahealth/httpclient.go [hunk 1]
```diff
@@ -172,8 +178,15 @@ func (h *HTTPClient) request(ctx context.Context, method, path string, body io.R
 		h.requestLock.Unlock()
 
 		if errors.Is(err, ratelimiter.ErrRateExceeded) {
-			time.Sleep(retryAfter)
-			return h.request(ctx, method, path, body, headers, out)
+			select {
+			case <-ctx.Done():
+				return nil, fmt.Errorf("waiting for rate limit timeout: %w", ctx.Err())
```
Comment (elliehastings) [inline athenahealth/httpclient.go:183]: wonder if there's a less ambiguous way to say this (like `"timed out waiting for rate limit retry interval"?) but totally fine as-is

**Review by elliehastings — APPROVED** [review]
Makes sense to me!

File: athenahealth/httpclient.go [hunk 2]
```diff
@@ -230,6 +243,10 @@ func (h *HTTPClient) request(ctx context.Context, method, path string, body io.R
 
 	res, err := h.httpClient.Do(req)
 	if err != nil {
+		if v := ctx.Value(ctxKeyRateLimitEncountered); v != nil {
+			return res, fmt.Errorf("prior rate limit encountered: %w", err)
```
[the remark this answers is not in the document]
Comment (dillonstreator) [reply athenahealth/httpclient.go:247]: Correct - this is just to indicate that there was a prior rate limit event encountered which may or may not be relevant to the actual error that was encountered
Not sure if there is a better way to handle this

File: athenahealth/httpclient.go (hunk 1 above)
Comment (dillonstreator) [reply athenahealth/httpclient.go:183]: In the current state, the full error would read something along the lines of:
`waiting for rate limit timeout: context deadline exceeded`
OR
`waiting for rate limit timeout: context canceled`
We can't assume that it was a timeout event as it's possible the context was canceled and not a deadline

I like the verbiage "retry interval" better so we can switch to
"waiting for rate limit retry interval"
which would give us:
`waiting for rate limit retry interval: context deadline exceeded`
OR
`waiting for rate limit retry interval: context canceled`

File: athenahealth/httpclient.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (willfitze) [reply athenahealth/httpclient.go:247]: Yea when we wrap errors it's to provide context for the error itself. In this case I don't think the error is actually ever related to rate limiting? If we just want to make note of the fact that we hit a rate limit does it make more sense to just log an info line in the `if errors.Is(err, ratelimiter.ErrRateExceeded)` block, maybe with the method and path as context?

[the remark this answers is not in the document]
Comment (dillonstreator) [reply athenahealth/httpclient.go]: Yea we probably do just want a log
I think this commit is a bit more appropriate https://github.com/eleanorhealth/go-athenahealth/pull/84/commits/aeae78257b6d065e6df3b0591b3303b0be6fa700

File: athenahealth/httpclient.go (hunk 1 above)
Comment (elliehastings) [reply athenahealth/httpclient.go]: Ah yup because `Done()` would catch cancellation as well as context timeouts 👍

**Review by elliehastings — APPROVED** [review]

**Pull request merged**